### PR TITLE
Fix ContractUpdateSuite and address ContractUpdateHandle comments

### LINKED ISSUE #8996: Fix ContractUpdateSuite tests
Waiting for implementation to be done in 
- https://github.com/hashgraph/hedera-services/issues/8623

### LINKED ISSUE #9654: Address `ContractUpdateHandler` review comments
### Problem

The very nicely executed [PR](https://github.com/hashgraph/hedera-services/pull/9379) introducing `ContractUpdateHandler` did have a few items left to do:
  - [x] We need to use `TokenServiceApi` to validate staking elections.
  - [x] We need to enforce the feature flag for contract auto-associations.
  - [x] We need to include the `hasMaxAutoAssociations()` value in the check for non-expiration field changes.
  - [x] We need to add unit tests.

### Solution

Address the above.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
Adding additional validations and unit tests

**Related issue(s)**:

Fixes #9654 
Fixes #8996

**Notes for reviewer**:
**Node:** Since I did some refactoring the changes made for this PR are not so clear - I will annotate them with Github comments.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java [hunk 1]
```diff
@@ -102,12 +105,112 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
+                .assertValidStakingElectionForUpdate(
+                        context.configuration()
+                                .getConfigData(StakingConfig.class)
+                                .isEnabled(),
+                        contract.declineReward(),
+                        contract.stakedId().kind().name(),
+                        contract.stakedAccountId(),
+                        contract.stakedNodeId(),
+                        accountStore,
+                        context.networkInfo());
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java:176]: New validation

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java [hunk 2]
```diff
@@ -102,12 +105,112 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
+            validateFalse(
+                    newMax > ledgerConfig.maxAutoAssociations(),
+                    REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
+
+            validateFalse(newMax < contract.maxAutoAssociations(), EXISTING_AUTOMATIC_ASSOCIATIONS_EXCEED_GIVEN_LIMIT);
+            validateFalse(
+                    entitiesConfig.limitTokenAssociations() && newMax > tokensConfig.maxPerAccount(),
+                    REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
+
+            validateTrue(contractsConfig.allowAutoAssociations(), NOT_SUPPORTED);
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java:154]: New validation

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java [hunk 3]
```diff
@@ -102,12 +105,112 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
+        return keyIsNotValid || candidate.contractID() != null;
+    }
+
+    private boolean onlyAffectsExpiry(ContractUpdateTransactionBody op) {
+        return !(op.hasProxyAccountID()
+                        || op.hasFileID()
+                        || affectsMemo(op)
+                        || op.hasAutoRenewPeriod()
+                        || op.hasAdminKey())
+                || op.hasMaxAutomaticTokenAssociations();
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java:197]: A new change

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java [hunk 4]
```diff
@@ -102,12 +105,112 @@ public void handle(@NonNull final HandleContext context) throws HandleException
...
+        final var currentMetadata =
+                new ExpiryMeta(contract.expirationSecond(), contract.autoRenewSeconds(), contract.autoRenewAccountId());
+        final var updateMeta = new ExpiryMeta(
+                op.hasExpirationTime() ? op.expirationTime().seconds() : NA,
+                op.hasAutoRenewPeriod() ? op.autoRenewPeriod().seconds() : NA,
+                null);
+        context.expiryValidator().resolveUpdateAttempt(currentMetadata, updateMeta);
+
+        context.serviceApi(TokenServiceApi.class)
+                .assertValidStakingElectionForUpdate(
```
Comment (vtronkov) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java:167]: @tinker-michaelj regarding this comment - https://github.com/hashgraph/hedera-services/pull/9379#discussion_r1381722155 :
The `assertValidStakingElection` was calling `validateStakedIdForCreation`. I see we have `validateStakedIdForUpdate` and I think we should call that instead. What do you think?

**Review by tinker-michaelj — APPROVED** [review]
This looks great, tyvm @vtronkov !

**Pull request merged**